### PR TITLE
Use imagec fetcher for login [full ci]

### PR DESCRIPTION
**Pull request by sflxn**
Refactored fetcher to separate auth and layer fetching. Moved
it to pkg/registry.  Updated the personality server's login to
use the fetcher code for login to match up the login and pull
auth code.

### CODE REVIEW

File: cmd/imagec/imagec.go [hunk 1]
```diff
@@ -200,23 +198,23 @@ func DestinationDirectory() string {
...
 		        └── library
 		            └── golang
 		                └── latest
 		                    ...
 		                    ├── f61ebe2817bb4e6a7f0a4cf249a5316223f7ecc886feac24b9887a490feaed57
-		                    │   ├── f61ebe2817bb4e6a7f0a4cf249a5316223f7ecc886feac24b9887a490feaed57.json
-		                    │   └── f61ebe2817bb4e6a7f0a4cf249a5316223f7ecc886feac24b9887a490feaed57.tar
+		                    │   ├── f61ebe2817bb4e6a7f0a4cf249a5316223f7ecc886feac24b9887a490feaed57.json
+		                    │   └── f61ebe2817bb4e6a7f0a4cf249a5316223f7ecc886feac24b9887a490feaed57.tar
 		                    └── manifest.json
```
Comment (caglar10ur) [inline cmd/imagec/imagec.go]: not sure what changed in this function, is it just indentation?

File: lib/apiservers/engine/backends/system.go [hunk 2]
```diff
@@ -242,7 +245,77 @@ func (s *System) UnsubscribeFromEvents(chan interface{}) {
...
+		Username: authConfig.Username,
+		Password: authConfig.Password,
+	})
+
+	// Only look at V2 registries
+	var registryAddress string
+	if strings.Contains(authConfig.ServerAddress, "/v2") {
+		registryAddress = authConfig.ServerAddress
+	} else {
+		registryAddress = authConfig.ServerAddress + "/v2/"
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: maybe?

```
registryAddress := authConfig.ServerAddress
if ! strings.Contains(authConfig.ServerAddress, "/v2") {
	registryAddress = registryAddress + "/v2/"
}
```

File: lib/apiservers/engine/backends/system.go [hunk 3]
```diff
@@ -242,7 +245,77 @@ func (s *System) UnsubscribeFromEvents(chan interface{}) {
...
+	}
+
+	registryUrl, err := url.Parse(registryAddress)
+	if err != nil {
+		msg := fmt.Sprintf("Bad login address: %s", registryAddress)
+		log.Debugf(msg)
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: s/Debugf/Errorf/g

File: lib/apiservers/engine/backends/system.go [hunk 4]
```diff
@@ -242,7 +245,77 @@ func (s *System) UnsubscribeFromEvents(chan interface{}) {
...
+			log.Debugf("Looking up OAuth URL failed: %s", err)
+			return "", err
+		}
+
+		log.Debugf("logging onto %s", authUrl.String())
+
+		// Just check if we get a token back.
+		token, err := fetcher.FetchAuthToken(authUrl)
+		if err != nil || token.Token == "" {
+			log.Debugf("Fetch auth token failed: %s", err)
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: s/Debugf/Errorf/g

File: lib/apiservers/engine/backends/system.go [hunk 5]
```diff
@@ -242,7 +245,77 @@ func (s *System) UnsubscribeFromEvents(chan interface{}) {
...
+
+		var authUrl *url.URL
+
+		// Attempt to get the Auth URL from HEAD operation to the registry
+		hdr, err := fetcher.Head(registryUrl)
+		if err == nil && fetcher.IsStatusUnauthorized() {
+			authUrl, err = fetcher.ExtractOAuthUrl(hdr.Get("www-authenticate"), nil)
+		}
+		if err != nil {
+			log.Debugf("Looking up OAuth URL failed: %s", err)
```
Comment (caglar10ur) [inline lib/apiservers/engine/backends/system.go]: s/Debugf/Errorf/g

File: pkg/fetcher/fetcher.go [hunk 6]
```diff
@@ -170,11 +180,36 @@ func (u *URLFetcher) Fetch(url *url.URL, ids ...string) (string, error) {
...
 }
 
-// fetch fetches the given URL using ctxhttp. It also streams back the progress bar only when ID is not an empty string.
-func (u *URLFetcher) fetch(ctx context.Context, url *url.URL, ID string) (string, error) {
+func (u *URLFetcher) FetchAuthToken(url *url.URL) (*Token, error) {
+	defer trace.End(trace.Begin(url.String()))
+
+	data, err := u.Fetch(url, false, nil)
+	if err != nil {
+		log.Debugf("Download failed: %v", err)
```
Comment (caglar10ur) [inline pkg/fetcher/fetcher.go]: s/Debugf/Errorf/g

File: pkg/fetcher/fetcher.go [hunk 7]
```diff
@@ -170,11 +180,36 @@ func (u *URLFetcher) Fetch(url *url.URL, ids ...string) (string, error) {
...
+		return nil, err
+	}
+
+	token := &Token{}
+
+	err = json.Unmarshal([]byte(data), &token)
+	if err != nil {
+		log.Debugf("Incorrect token format: %v", err)
```
Comment (caglar10ur) [inline pkg/fetcher/fetcher.go]: s/Debugf/Errorf/g

**Comment by caglar10ur** [comment]
minor comments otherwise lgtm

File: lib/apiservers/engine/backends/system.go (hunk 2 above)
Comment (anchal-agrawal) [inline lib/apiservers/engine/backends/system.go]: Would `strings.HasSuffix` be more suitable? Not that it makes a difference in this case, I think.

**Comment by mdubya66** [comment]
This thing has been open for 17 days. Is it still valid? If so, then it will need [full ci] before merged

**Comment by hickeng** [comment]
LGMT
@jzt does this duplicate anything you're doing with the imagec lib work?
@mdubya66 still looks valid unless superseded by@jzt work.

**Comment by jzt** [comment]
@hickeng @mdubya66 no, it doesn't appear to duplicate any of my work. I will need to rebase against this, though, once it's in.

File: pkg/fetcher/fetcher.go [hunk 8]
```diff
@@ -245,7 +294,7 @@ func (u *URLFetcher) fetch(ctx context.Context, url *url.URL, ID string) (string
 	defer out.Close()
 
 	// Stream into it
-	_, err = io.Copy(out, in)
+	_, err = io.Copy(out, rdr)
```
Comment (vburenin) [inline pkg/fetcher/fetcher.go]: it there no interest to log an error and number of copied bytes?

**Comment by mdubya66** [comment]
lgtm

File: cmd/imagec/imagec.go (hunk 1 above)
Comment (sflxn) [inline cmd/imagec/imagec.go]: Yeah, must have been my editor.

File: lib/apiservers/engine/backends/system.go (hunk 2 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: sounds good.

File: lib/apiservers/engine/backends/system.go (hunk 3 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/system.go]: changed.

File: pkg/fetcher/fetcher.go (hunk 6 above)
Comment (sflxn) [inline pkg/fetcher/fetcher.go]: changed.

File: pkg/fetcher/fetcher.go (hunk 7 above)
Comment (sflxn) [inline pkg/fetcher/fetcher.go]: changed.

**Comment by sflxn** [comment]
ping

**Comment by sflxn** [comment]
ping

**Pull request merged**